docs/proposals/20210222-azure-machinepool-machine.md

```diff
@@ -0,0 +1,332 @@
+---
+title: Azure Machine Pool Machines
+authors:
+- @devigned
+  reviewers:
+- "@CecileRobertMichon"
+- @nader-ziada
+  creation-date: 2021-02-22
+  last-updated: 2021-02-22
+  status: implementable
+  see-also:
+- https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/819
+- https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1067
+---
+
+
+# Azure Machine Pool Machines
+
+## Table of Contents
+- [Summary](#summary)
+- [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals / Future Work](#non-goals--future-work)
+- [Proposal](#proposal)
+    - [User Stories](#user-stories)
+        - [Story 1 - Upgrading the Kubernetes Version of a MachinePool](#story-1---upgrading-the-kubernetes-version-of-a-machinepool)
+        - [Story 2 - Reducing the Number of Replicas in a MachinePool](#story-2---reducing-the-number-of-replicas-in-a-machinepool)
+    - [Requirements](#requirements)
+        - [Functional](#functional)
+        - [Non-Functional](#non-functional)
+    - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+        - [Proposed API Changes](#proposed-api-changes)
+        - [Proposed Controllers Changes](#proposed-controller-changes)
+        - [Proposed Changes of Responsibily](#proposed-changes-of-responsibility)
+- [Available Options](#available-options-for-cluster-api-provider-azure)
+    - [Add Annotations to AzureMachinePool for Instance Delete Selection](#option-1-add-annotations-to-azuremachinepool-for-instance-delete-selection)
+        - [Pros](#option-1-pros)
+        - [Cons](#option-1-cons)
+    - [Separate AzureMachinePool and AzureMachinePoolMachines](#option-2-separate-azuremachinepool-and-azuremachinepoolmachines)
+        - [Pros](#option-2-pros)
+        - [Cons](#option-2-cons)
+- [Conclusions](#conclusions)
+- [Additional Details](#additional-details)
+    - [Test Plan](#test-plan)
+- [Implementation History](#implementation-history)    
+
+## Summary
+
+Azure MachinePool currently embeds the state of each of the machines in the MachinePool within the status of the Azure
+MachinePool. MachinePool machines should be their own resources to enable individual lifecycles.
+
+## Motivation
+
+By giving each AzureMachinePoolMachine an individual lifecycle, a user would be able to inform CAPZ of the specific
+instance to delete and then have the AzureMachinePoolMachine controller cordon and drain the node prior to deleting
+the underlying infrastructure.
+
+Azure Virtual Machine Scale Set Instances do not provide the specific machine events needed to alert Kubernetes when
```

Comment:
Can't we leverage the [Azure scheduled events](https://docs.microsoft.com/en-us/azure/virtual-machine-scale-sets/virtual-machine-scale-sets-terminate-notification)?

Reply:
> Terminate notifications only on ‘delete’ operations – All delete operations (manual delete or Autoscale-initiated scale-in) will generate Terminate events if your scale set has scheduledEventsProfile enabled. Other operations such as reboot, reimage, redeploy, and stop/deallocate do not generate Terminate events. 

We could if we decide to only delete and replace machines, which will cause infra to deprovision and then provision again. Another way we could do this is to use `Reimage`, which does not deprovision the infra, but only updates the model. In `Reimage`, we would not get an event from Azure letting the machine know that it should cordon and drain.

On the other hand, if we are using spot instances, we will likely need to listen to machine events to know when a node must cordon and drain.

This thread covers the problem pretty well: #819. I do think we will need some on host component like https://github.com/awesomenix/drainsafe at some point to help with spot instances or VM migration.

Reply:
Oh I see. Thanks for the clarification. Still, I'd reword it a bit because it seems VMSS do provide machine events when an instance is deleted.

Reply:
@fiunchinho wdyt of the termination notifications wording now?

Reply:
Perfect.